### PR TITLE
modify thrift output folder to prevent clashing with other plugins

### PR DESCRIPTION
**Pull request by joprice**
The current default for `scroogeThriftOutputFolder` causes compilation errors when used alongside other sbt plugins that generate source, the most common being sbt-buildinfo. This solves it by generating the code in a subfolder of the generated sources folder. This also adds scripted setup to test the sbt plugin. The test can be run with `scrooge-sbt-plugin/scripted`.

### CODE REVIEW

**Comment by kevinoliver** [comment]
i'm pretty weak in terms of sbt knowledge so apologies in advance for my questions on the diff.

File: project/Build.scala [hunk 1]
```diff
@@ -278,7 +280,9 @@ object Scrooge extends Build {
       publishMavenStyle := false,
       repository in bintray := "sbt-plugins",
       licenses += (("Apache-2.0", url("http://www.apache.org/licenses/LICENSE-2.0.html"))),
-      bintrayOrganization in bintray := Some("twittercsl")
+      bintrayOrganization in bintray := Some("twittercsl"),
+      scriptedBufferLog := false,
```
Comment (kevinoliver) [inline project/Build.scala]: can you leave comments documenting what these options mean

File: project/plugins.sbt [hunk 2]
```diff
@@ -14,3 +14,6 @@ addSbtPlugin("net.virtual-void" % "sbt-dependency-graph" % "0.8.2")
 addSbtPlugin("org.scoverage" % "sbt-coveralls" % "1.0.0")
 addSbtPlugin("org.scoverage" % "sbt-scoverage" % "1.2.0")
 addSbtPlugin("pl.project13.scala" % "sbt-jmh" % "0.2.2")
+
+libraryDependencies += "org.scala-sbt" % "scripted-plugin" % sbtVersion.value
```
Comment (kevinoliver) [inline project/plugins.sbt]: am i correct that this is separate from the fix and enables testing of plugins? 

if so, can we split this into to two separate PRs?

File: scrooge-sbt-plugin/src/sbt-test/scrooge/basic/build.sbt [hunk 3]
```diff
@@ -0,0 +1,10 @@
+
+scalaVersion := "2.11.8"
```
Comment (kevinoliver) [inline scrooge-sbt-plugin/src/sbt-test/scrooge/basic/build.sbt]: i thought sbt plugins were 2.10 only?

File: scrooge-sbt-plugin/src/sbt-test/scrooge/basic/build.sbt [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+
+enablePlugins(BuildInfoPlugin)
+
+libraryDependencies ++= Seq(
```
Comment (kevinoliver) [inline scrooge-sbt-plugin/src/sbt-test/scrooge/basic/build.sbt]: these versions not correct when we are doing development. can we share the code thats in `Build.scala` for setting these?

File: scrooge-sbt-plugin/src/sbt-test/scrooge/basic/test [hunk 5]
```diff
@@ -0,0 +1,3 @@
+> compile
+$ exists target/scala-2.11/src_managed/main/thrift/com/twitter/demo/User.scala
+$ exists target/scala-2.11/src_managed/main/sbt-buildinfo/BuildInfo.scala
```
Comment (kevinoliver) [inline scrooge-sbt-plugin/src/sbt-test/scrooge/basic/test]: this is pretty cool.

File: scrooge-sbt-plugin/src/sbt-test/scrooge/basic/build.sbt (hunk 3 above)
Comment (joprice) [inline scrooge-sbt-plugin/src/sbt-test/scrooge/basic/build.sbt]: This version is the version of a test project using this plugin.

File: project/plugins.sbt (hunk 2 above)
Comment (joprice) [inline project/plugins.sbt]: Yes this enables testing of the plugin, but the test I added is for this specific fix. You want me to separate that into another pr with a project that doesn't test this issue first, just a simple test that the plugin generates a scala file?

File: scrooge-sbt-plugin/src/sbt-test/scrooge/basic/build.sbt (hunk 4 above)
Comment (joprice) [inline scrooge-sbt-plugin/src/sbt-test/scrooge/basic/build.sbt]: Yea I was hoping for some direction on what to do with these. I thought it might be ok if the plugin is backwards compatible, but if that is not the case, we could export some more props for these versions.d

Comment (kevinoliver) [inline scrooge-sbt-plugin/src/sbt-test/scrooge/basic/build.sbt]: yeah, i think we need it to work the same as it does in `Build.scala`.

File: scrooge-sbt-plugin/src/sbt-test/scrooge/basic/build.sbt (hunk 3 above)
Comment (kevinoliver) [inline scrooge-sbt-plugin/src/sbt-test/scrooge/basic/build.sbt]: got it, thanks.

File: project/plugins.sbt (hunk 2 above)
Comment (kevinoliver) [inline project/plugins.sbt]: i see. ok lets leave them bundled.

File: project/Build.scala (hunk 1 above)
Comment (joprice) [inline project/Build.scala]: The documentation for scripted is here: http://www.scala-sbt.org/0.13/docs/Testing-sbt-plugins.html. Do you want me to just put that link in a comment?

**Comment by joprice** [comment]
@kevinoliver I updated the scripted test to use `sys.props` for dependency versions.

File: project/Build.scala (hunk 1 above)
Comment (kevinoliver) [inline project/Build.scala]: that'd be awesome. thanks.

**Comment by kevinoliver** [comment]
mind adding that one doc link, and then this lgtm. thanks @joprice

**Comment by joprice** [comment]
Added the doc link. One more thing to note is that you have to run `scrooge-sbt-plugin/scripted` to kick off the tests. Should that be added to `.travis.yml`? It seems like the travis builds have been failing for a while.

**Comment by kevinoliver** [comment]
yeah please add it to `.travis.yml`, thanks. 

i /think/ travis is back to working now, we had some sbt dependency issues we were working through.

**Comment by joprice** [comment]
It only makes sense to run `scripted` on travis for scala 2.10, since it's an sbt plugin. So I would add it to `script` in `.travis.yml` without `++$TRAVIS_SCALA_VERSION` and it will get run twice. I haven't used it, but it looks like we could make use of the 'matrix' configuration and have a different command for each version: https://github.com/twitter/summingbird/blob/master/.travis.yml#L12

**Comment by mosesn** [comment]
I think doing the matrix style or the "run the same thing twice" are both OK.  As long as it doesn't get in our way, I'm happy.

File: .travis.yml [hunk 6]
```diff
@@ -34,5 +34,14 @@ before_script:
   - ./bin/travisci
   - travis_retry ./sbt ++$TRAVIS_SCALA_VERSION update
 
-script: ./sbt ++$TRAVIS_SCALA_VERSION coverage test && ./sbt ++$TRAVIS_SCALA_VERSION coverageAggregate
+script: "echo no op"
```
Comment (cacoco) [inline .travis.yml]: I think your diff is out of date. We've switched to use codecov.io in the latest code (instead of coveralls). You may want to re-base the branch.

**Comment by joprice** [comment]
Thanks! I'll update.

**Comment by mosesn** [comment]
@joprice this looks good to me!  I'm going to try to resolve the merge conflicts and merge this in.  Thanks for the PR!

**Comment by joprice** [comment]
Thanks! Unfortunately, I wasn't able to get this to pass CI :-(

**Comment by mosesn** [comment]
@joprice no worries, we've had a lot of CI problems, so I'll see what I can do on our end.

**Comment by mosesn** [comment]
@joprice I might be running into the same thing you did when trying to get CI to work.  was it problems around finding a copy of sbt 0.13.9 for the test?

**Comment by joprice** [comment]
I don't remember the exact error sorry.

**Comment by mosesn** [comment]
@joprice OK, this unfortunately looks like it's going to be something of a doozy because of the way scripted works, and it doesn't look like the scripted maintainers are very [responsive][0].  I'm going on vacation for a week, and will try again when I come back, but we might have to rip the test out of it to merge it in until we can figure out how to get scripted to work in our environment.

[0]: https://github.com/sbt/sbt/issues/2685

**Comment by mosesn** [comment]
@joprice I think getting scripted to work with our stuff is going to be more trouble than it's worth for now.  I think I'm going to rip out the testing and then merge it in, this has taken way too long.  Sorry, I'll work in the future on getting scripted working.

**Comment by joprice** [comment]
No problem. I just enabled the new 'allow edits from maintainers'.

**Comment by jcrossley** [comment]
Hey @joprice, wanted to give you an update on this. Unfortunately, we can't merge it in yet (even without the scripted plugin and tests), because it's causing issues with the [Finatra ](https://github.com/twitter/finatra) framework. Trying to sort this out and will update you with progress. Thanks for contributing!

File: scrooge-sbt-plugin/src/main/scala/com/twitter/ScroogeSBT.scala [hunk 7]
```diff
@@ -171,7 +171,7 @@ object ScroogeSBT extends AutoPlugin {
     scroogeDefaultJavaNamespace := "thrift",
     scroogeThriftSourceFolder <<= (sourceDirectory) { _ / "thrift" },
     scroogeThriftExternalSourceFolder <<= (target) { _ / "thrift_external" },
-    scroogeThriftOutputFolder <<= (sourceManaged) { identity },
+    scroogeThriftOutputFolder in Compile := (sourceManaged in Compile).value / "thrift",
```
Comment (cacoco) [inline scrooge-sbt-plugin/src/main/scala/com/twitter/ScroogeSBT.scala]: As the comments above this section point out the configuration here should be in both Compile *and* test configuration. Hence this is a breaking change since it sets the output folder in the Compile configuration only. This probably needs to be modified to be:

```
scroogeThriftOutputFolder in Compile := (sourceManaged in Compile).value / "thrift",
scroogeThriftOutputFolder in Test := (sourceManaged in Test).value / "thrift",
```

**Comment by cacoco** [comment]
@joprice we're going to apply the change internally and get this merged as soon as we can. The change is under review internally. Thanks!

**Comment by cacoco** [comment]
@joprice  this has been merged internally and should show up upon the next sync to Github. Thanks!

**Comment by cacoco** [comment]
Merged in e604a7212e5eb73445a47a36223ea9f26e419607
